src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -175,9 +183,17 @@ private void RestorePackages()
             catch (Exception ex)
             {
                 TraceWriter.Error(ex.Message, ex is CompilationErrorException ? null : ex);
+
+                startedEvent.Success = false;
+                TraceWriter.Error(ex.Message, ex);
+
                 TraceWriter.Verbose("Function completed (Failure)");
                 throw;
             }
+            finally
+            {
+                _metrics.EndEvent(startedEvent);
+            }
         }
```

Comment:
Would it make sense to make Invoke non-virtual, introduce a protected abstract that derived types would use to perform their logic (e.g. ProcessInvoke, ProcessInvokeAsync...) and move the metrics code to the base, wrapping the call to that abstract method?